### PR TITLE
feat(postgres): Support for ARRAY(ENUM) datatype.

### PR DESCRIPTION
**Pull request by wavded**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Did you follow the commit message conventions explained in CONTRIBUTING.md?

### Description of change

This commit addresses the issue raised in #1498.  I welcome feedback as I'm new to this code base.  After trying a few different ways to get this to work, this seemed the least invasive to me.

### CODE REVIEW

File: lib/query-interface.js [hunk 1]
```diff
@@ -189,12 +189,21 @@ class QueryInterface {
...
         const promises = [];
         let enumIdx = 0;
+        let type;
 
         for (i = 0; i < keyLen; i++) {
-          if (attributes[keys[i]].type instanceof DataTypes.ENUM) {
+          type = attributes[keys[i]].type;
+          // Handle arrays of enum type.
+          if (type instanceof DataTypes.ARRAY) type = type.type;
+          // } else if (definition.type instanceof DataTypes.RANGE || DataTypes.ARRAY.is(definition.type, DataTypes.RANGE)) {
```
Comment (wavded) [inline lib/query-interface.js]: I will remove this line.  Just an artifact.

**Comment by wavded** [comment]
The tests passed the first time this PR built, then I removed a comment, and Travis failed, seems unrelated.  Seems other PRs are suffering similar issues.

File: lib/model.js [hunk 2]
```diff
@@ -974,6 +974,14 @@ class Model {
         this._dataTypeChanges[name] = definition.type._isChanged;
       }
 
+      // Mark array enum value for Postgres.
+      if (definition.type instanceof DataTypes.ARRAY &&
+        DataTypes.ARRAY.is(definition.type, DataTypes.ENUM)) {
+        definition.type.enumName = '"' + (this._schema || 'public')
```
Comment (sushantdhiman) [inline lib/model.js]: Use template strings for better readability

File: lib/query-interface.js [hunk 3]
```diff
@@ -189,12 +189,20 @@ class QueryInterface {
...
 
         for (i = 0; i < keyLen; i++) {
-          if (attributes[keys[i]].type instanceof DataTypes.ENUM) {
+          type = attributes[keys[i]].type;
+          // Handle arrays of enum type.
+          if (type instanceof DataTypes.ARRAY) type = type.type;
+
+          const attr = attributes[keys[i]].field || keys[i];
+          if (type instanceof DataTypes.ENUM) {
+            type._enumName = this.QueryGenerator.pgEnumName(tableName, attr, { schema: false });
```
Comment (sushantdhiman) [inline lib/query-interface.js]: Why you need to set this, `_enumName` used anywhere else or it should be `enumName` ?

File: lib/model.js (hunk 2 above)
Comment (wavded) [reply lib/model.js]: Will do, wasn't sure if you wanted to keep older Node compatibility or not.

File: lib/query-interface.js (hunk 3 above)
Comment (wavded) [reply lib/query-interface.js]: Good catch, this isn't needed to be set, will clean that up.

Comment (wavded) [reply lib/query-interface.js]: Was left over code from a previous approach I had tried that I forgot to remove.

**Comment by sushantdhiman** [comment]
Approach looks good. I have but one concern, we are now building enum name ourselves in current PR. When enum is may be user defined(unknown to sequelize) or name is different (or changed in future) how we are going to handle that.

May be at sequelize end we can use single method, may be from in ENUM type to get its name. Also allow users to override it if they want to use some other enum.

Also we need to integration tests as I mentioned earlier.

**Comment by wavded** [comment]
@sushantdhiman Good points.  In addition to the enum name, I also discovered we need the oids for the custom enum types for the PG driver to correctly parse the array).  Any recommendations on where would be a good spot to hook into the initialization process to grab these details? (agree that a single method would be nice)

**Comment by sushantdhiman** [comment]
I got a chance to work on `ARRAY(ENUM)` support. It will be added by https://github.com/sequelize/sequelize/pull/8738

Closing this PR, Thanks for this initiation :)

**Comment by wavded** [comment]
@sushantdhiman much appreciated!!

**Pull request closed without merging**